### ISSUE TITLE
Set number of ranks in `mpi4py.futures.MPIPoolExecutor`

### ISSUE THREAD

**Issue by jan-janssen** [mpi4py/mpi4py#324]
I would like to adjust the number of MPI tasks which are assigned to one process in the `MPIPoolExecutor`. Currently I use `MPI.COMM_SELF` to access the MPI tasks of a given process in an `MPIPoolExecutor`, still it is just a single `MPI` task per process, but I would like to control the number of `MPI` tasks per process. Is this possible?

**Comment by dalcinl** [comment]
Sorry, I do not quite understand what you are asking for. What do you mean by "MPI task"? Why you put "MPI" in front of "task"? Do you mean "parallel" tasks?

Let me tell you in advance that `mpi4py.futures` primary motivation is to implement the Python stdlib `concurrent.futures` interface as verbatim as possible to allow it scale to multiple compute nodes. In some sense, `mpi4py.futures` tries hard to "hide" MPI from the user, you need almost zero MPI knowledge to profit from the package. 
A consequence of this design is that tasks are sequential. if you are trying to use `mpi4py.futures` with inter-task coordination and MPI communication, then that will simply not work, you have absolutely no control on where a task is executed nor any ability for inter-task coordination. Implementing such a feature set would be way beyond the simple interface of `concurrent.futures` and it probably complicated enough to be part of a separate project.

**Comment by jan-janssen** [comment]
Maybe should have said `ranks` rather than processes. I have an atomistic simulation code written in C++ with python bindings an mpi4py support. I can either run one large simulation with all MPI ranks being used, or I can provide the `MPI.COMM_SELF` communicator inside an `MPIPoolExecutor` which allows me to run multiple calculation in parallel, independent from each other. Now I want to use an intermediate setup, with 4 parallel calculation with each calculation using 4 mpi ranks. Is this possible with the `MPIPoolExecutor`?

**Comment by dalcinl** [comment]
I see what you mean now. And as per my previous reply the answer to your question would be no. However, I was wrong, now I realize it is actually quite possible to run parallel coordinated tasks by using a trick and following a simple rule. 

Let's use as an example the following parallel *Hello World!* using `mpi4py.futures`.
```python
from mpi4py import MPI
from mpi4py.futures import MPIPoolExecutor, wait

def helloworld():
    MPI.COMM_WORLD.Barrier()
    size = MPI.COMM_WORLD.Get_size()
    rank = MPI.COMM_WORLD.Get_rank()
    name = MPI.Get_processor_name()
    return f"Hello, World! I am worker {rank} of {size} on {name}."

if __name__ == '__main__':
    executor = MPIPoolExecutor(5)
    num_workers = executor._max_workers
    futures = []
    for _ in range(num_workers):
        f = executor.submit(helloworld)
        futures.append(f)
    wait(futures)
    for f in futures:
        print(f.result())
    executor.shutdown()
```
Now execute it with just one MPI process as usual, the executor will use 5 workers.
```
$ mpiexec -n 1 python helloworld.py
Hello, World! I am worker 0 of 5 on localhost.
Hello, World! I am worker 1 of 5 on localhost.
Hello, World! I am worker 2 of 5 on localhost.
Hello, World! I am worker 3 of 5 on localhost.
Hello, World! I am worker 4 of 5 on localhost.
```

From now on I'm assuming your executor is actually using `Spawn()` (as above), that is,  the script is not being executed with `python -m mpi4py.futures ...` in legacy mode as explained in the docs. For legacy mode, things could still be worked out, but it would be better if I add some documented little extra feature to mpi4py.futures, that is, access to an intracommunicator that is guaranteed to contain the workers and just the workers. In the case of spawned workers, `COMM_WORLD` is such communicator.

Now the gory details why the approach is guaranteed to work. 

* Notice I'm getting the number of workers with `executors._max_workers`. Then I submit EXACTLY one task per worker, no more, no less. Each returned `future` is collected in a list.
* Notice the use of `COMM_WORLD` in the `helloworld()` function. This world communicator is the one corresponding to all the workers, the parent or master process (i.e the one that submits tasks) is not included in it.
* Notice also the `Barrier()` call at the very beginning of `helloworld`. As a consequence of that barrier, and the fact that you are submitting exactly one task per worker, as you submit the tasks, they block in the barrier until the last one start executing and enters the barrier. This is key to prevent that let say the first worker executes a task quickly enough for a new task being assigned to it.
* The `wait()` call for all the futures is optional. In this particular example you do not really need it for the code to work correctly. But maybe in some more convoluted scenarios you may want to be certain that ranks finished before the master execution continues.

Now you have a away to somehow run run a single parallel task. 

But you cannot have enough, next you want to run multiple independent parallel tasks. There are two possible options, one easy, the other more involved.
1. The easy way is to create an executor for each parallel execution. Each executor can work independently. Each executor will spawn its own workers. That's all.
2. The harder way, if you want to brag about your deep MPI knowledge, is to create only one executor, and submit tasks exactly as in my example above, up to the point right after the `Barrier()` call. Afterwards, you take `COMM_WORLD` and you split it in sub communicators. The easiest way to do that in mpi4py would be to use the `Comm.Split()` method,  defining subgroups by computing a color that depends on the rank in `COMM_WORLD`. This is not the only way though: MPI also offers `Group` operations to define the subgroup and then use `Comm.Create[_group]()`. But I would argue that using `Comm.Split()` is just fine. After you created your `subcomm`, you use that communicator to perform your parallel computations. 

As you can see, approach (2) is more involved. Moreover, approach (2) require that your computational code does things properly the MPI way. That is, the code takes a communicator from the user and uses it for all communication, and NEVER EVER uses `MPI_COMM_WORLD` explicitly for communication. Library code should NEVER use `MPI_COMM_WORLD` explicitly, if it does, then it is bad code. `MPI_COMM_WORLD` should be left for USERS of the library to pass it down. Otherwise, users like you would not be able to use such code in subcommunicators. You most likely know about these things, I'm just repeating them again for others that may end up reading this text and are not aware.

Let me know the outcome.

**Comment by jan-janssen** [comment]
@dalcinl It took me a while to go through this, the part that I do not understand is why you call MPI with just a single process `mpiexec -n 1`. I tried it with `mpiexec -n 5` but this obviously executed the same script 5 times. Can you explain this difference a bit more?

**Comment by dalcinl** [comment]
> It took me a while to go through this, the part that I do not understand is why you call MPI with just a single process `mpiexec -n 1`

The important line is `executor = MPIPoolExecutor(5)`, so th executor will use 5 workers spawned via MPI. You still have to use `mpiexec -n1` to spawn the "parent" or "master" process that creates uses the executor, but that requirement is related to the issue/limitation we discussed elsewhere and that should be fixed in MPICH 4.1+.

**Comment by jan-janssen** [comment]
@dalcinl I have it working with `openmpi` but I cannot get it to work with `mpich` or `msmpi`. With `openmpi` I had to use the `--oversubscribe` option, do you know if a similar feature is available for the other MPI implementations?

**Comment by jan-janssen** [comment]
> @dalcinl I have it working with `openmpi` but I cannot get it to work with `mpich` or `msmpi`. With `openmpi` I had to use the `--oversubscribe` option, do you know if a similar feature is available for the other MPI implementations?

I was able to solve it by just removing the `--oversubscribe`, somehow `mpich` and `msmpi` hang forever with this option and `openmpi` does without it. Now I set the option via the environment. By the way, is there a way to see which MPI version is calling `mpi4py` from the python side? Something like `MPI.distribution` which then returns `openmpi`, `mpich` or `msmpi` ?

**Comment by dalcinl** [comment]
What’s the issue with MPICH?

If you have to enable oversubscription with openmpi, that means you are using a total number of MPI processes used at the same time is larger than the number of cores. For compute-bound tasks that’s hardly a good idea. How are you distributing your work across cores?

**Comment by dalcinl** [comment]
If you are really oversubscribed, then MPICH will behave badly. As I said, oversubscription is hardly useful for computational tasks. 

For MPI implementation info use:
https://mpi4py.readthedocs.io/en/stable/reference/mpi4py.MPI.Get_library_version.html#mpi4py.MPI.Get_library_version
This is a standard MPI functionality.

**Comment by dalcinl** [comment]
@jan-janssen FYI, MPICH 4.1.1 is now available in conda-forge. `MPIPoolExecutor` should work out of the box on Jupyter notebooks, although I have not tried.

PS: Can we close this issue?

**Comment by jan-janssen** [comment]
@dalcinl Sure feel free to close the issue and thanks a lot for all the support and your recent developments towards a multiprocessing pool like interface.

**Comment by jan-janssen** [comment]
@dalcinl I tend to come back to this example. Is there anyway around the MPI `Spawn()` call? Is it possible to reuse the existing MPI communicators? I would like to call `mpiexec -n 1 python helloworld.py` with `mpiexec -n 5 python helloworld.py` rather than setting the number of tasks in `MPIPoolExecutor(5)`.

**Comment by dalcinl** [comment]
I'm working on something in branch `feature/futures`, look at the example `demo/futures/run_hello.py`. This work is not yet ready, things get even more convoluted than they already are, and I'll not rush it until I'm happy.
